daemon/sailjailclient.c

```diff
@@ -0,0 +1,1379 @@
+            ++app_argv;
+        }
+
+        if( !want ) {
+            /* Template args exhausted */
+            if( *app_argv ) {
+                /* Excess application args */
+                log_err("argv has unwanted '%s'", *app_argv);
+                goto EXIT;
+            }
+            break;
+        }
+
+        int field_code = sailjailclient_get_field_code(want);
+
+        if( !field_code ) {
+            /* Exact match needed */
+            if( g_strcmp0(*app_argv, want) ) {
+                /* Application args has something else */
+                log_err("argv is missing '%s'", want);
+                goto EXIT;
+            }
+            ++app_argv;
+            continue;
+        }
+
+        /* Field code explanations from "Desktop Entry Specification"
+         *
+         * https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#exec-variables
+         */
+        int code_args = 0;
+        switch( field_code ) {
+        case 'f':  /* A single file name */
+        case 'u':  /* A single URL */
+        case 'c':  /* The translated name of the application */
+        case 'k':  /* The location of the desktop file */
+            code_args = 1;
+            break;
+        case 'F':  /* A list of files */
+        case 'U':  /* A list of URLs */
+            code_args = -1;
+            break;
+        case 'i':
+            /* The Icon key of the desktop entry expanded as two
+             * arguments, first --icon and then the value of the
+             * Icon key. Should not expand to any arguments if
+             * the Icon key is empty or missing.
+             */
+            if( !g_strcmp0(*app_argv, "--icon") )
+                ++app_argv, code_args = 1;
+            break;
+        case 'd':
+        case 'D':
+        case 'n':
+        case 'N':
+        case 'v':
+        case 'm':
+            /* Deprecated */
+            log_err("Exec line has deprecated field code '%s'", want);
+            goto EXIT;
+        default:
+            /* Unknown */
+            log_err("Exec line has unknown field code '%s'", want);
+            goto EXIT;
+        }
+
+        if( code_args < 0 ) {
+            /* Any number of args */
+            if( sailjailclient_get_field_code(*tpl_argv) ) {
+                log_err("Can't validate '%s %s' combination", want, *tpl_argv);
+                goto EXIT;
+            }
+            while( *app_argv && g_strcmp0(*app_argv, *tpl_argv) ) {
+                if( sailjailclient_is_option(*app_argv) ) {
+                    log_err("option '%s' at field code '%s'", *app_argv, want);
+                    goto EXIT;
+                }
+                ++app_argv;
+            }
+        }
+        else {
+            /* Specified number of args */
+            for( ; code_args > 0; --code_args ) {
+                if( !*app_argv ) {
+                    log_err("missing args for field code '%s'", want);
```

Comment:
I noticed that I hit this error when I tested with an app that specifies %f on its command line but I don't give any path for it to handle. I think that's not quite correct. This is what the specification says:

> If the application should not open any file the %f, %u, %F and %U field codes must be removed from the command line and ignored. 

I understand this so that fulfulling the field codes is optional.

Reply:
Should be fixed.